internal/fs/fs.go

```diff
@@ -2058,6 +2061,60 @@ func (fs *fileSystem) checkDirNotEmpty(dir inode.BucketOwnedDirInode, name strin
 	return nil
 }
 
+// Rename an old folder to a new folder in hierarchical bucket. If the new folder already
+// exists and is non-empty, return ENOTEMPTY.
+//
+// LOCKS_EXCLUDED(fs.mu)
+// LOCKS_EXCLUDED(oldParent)
+// LOCKS_EXCLUDED(newParent)
+func (fs *fileSystem) renameFolder(ctx context.Context, oldParent inode.DirInode, oldName string, newParent inode.DirInode, newName string) (err error) {
+	var pendingInodes []inode.DirInode
+	defer fs.releaseInodes(&pendingInodes)
+
+	oldDir, err := fs.getBucketDirInode(ctx, oldParent, oldName)
+	if err != nil {
+		return err
+	}
+	pendingInodes = append(pendingInodes, oldDir)
+
+	if err = fs.ensureNoOpenFilesInDirectory(oldDir, oldName); err != nil {
+		return err
+	}
+
+	newDir, err := fs.getBucketDirInode(ctx, newParent, newName)
+	if err == nil {
```

Comment:
What if the err is not nil? Does that indicate that everything is right for proceeding?

Reply:
If err is not nil means we don't get Inode, that refers directory does not exist. So we can go ahead with renaming